From 45deef4f3d27f0fcc1fd92781dadfe659ada7f70 Mon Sep 17 00:00:00 2001
From: Robin Gong <b38343@freescale.com>
Date: Wed, 7 Jan 2015 18:11:08 +0800
Subject: [PATCH] MLK-10244-3 dma: imx-sdma: add new i.mx7d support

Add new chip i.mx7d support,both ROM script and RAM script changed.

Signed-off-by: Robin Gong <b38343@freescale.com>
(cherry picked from commit 99b19e2f2a5f0450ab30794a3dd170129640c2c1)
---
 drivers/dma/imx-sdma.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/drivers/dma/imx-sdma.c b/drivers/dma/imx-sdma.c
index bae2ea8..7055c9e 100644
--- a/drivers/dma/imx-sdma.c
+++ b/drivers/dma/imx-sdma.c
@@ -457,6 +457,24 @@ static struct sdma_driver_data sdma_imx6sx = {
 	.script_addrs = &sdma_script_imx6sx,
 };
 
+static struct sdma_script_start_addrs sdma_script_imx7d = {
+	.ap_2_ap_addr = 644,
+	.uart_2_mcu_addr = 819,
+	.mcu_2_app_addr = 749,
+	.uartsh_2_mcu_addr = 1034,
+	.mcu_2_shp_addr = 962,
+	.app_2_mcu_addr = 685,
+	.shp_2_mcu_addr = 893,
+	.spdif_2_mcu_addr = 1102,
+	.mcu_2_spdif_addr = 1136,
+};
+
+static struct sdma_driver_data sdma_imx7d = {
+	.chnenbl0 = SDMA_CHNENBL0_IMX35,
+	.num_events = 48,
+	.script_addrs = &sdma_script_imx7d,
+};
+
 static struct platform_device_id sdma_devtypes[] = {
 	{
 		.name = "imx25-sdma",
@@ -480,6 +498,9 @@ static struct platform_device_id sdma_devtypes[] = {
 		.name = "imx6sx-sdma",
 		.driver_data = (unsigned long)&sdma_imx6sx,
 	}, {
+		.name = "imx7d-sdma",
+		.driver_data = (unsigned long)&sdma_imx7d,
+	}, {
 		/* sentinel */
 	}
 };
@@ -487,6 +508,7 @@ MODULE_DEVICE_TABLE(platform, sdma_devtypes);
 
 static const struct of_device_id sdma_dt_ids[] = {
 	{ .compatible = "fsl,imx6sx-sdma", .data = &sdma_imx6sx, },
+	{ .compatible = "fsl,imx7d-sdma", .data = &sdma_imx7d, },
 	{ .compatible = "fsl,imx6q-sdma", .data = &sdma_imx6q, },
 	{ .compatible = "fsl,imx53-sdma", .data = &sdma_imx53, },
 	{ .compatible = "fsl,imx51-sdma", .data = &sdma_imx51, },
@@ -1581,6 +1603,7 @@ static void sdma_issue_pending(struct dma_chan *chan)
 #define SDMA_SCRIPT_ADDRS_ARRAY_SIZE_V1	34
 #define SDMA_SCRIPT_ADDRS_ARRAY_SIZE_V2	38
 #define SDMA_SCRIPT_ADDRS_ARRAY_SIZE_V3	41
+#define SDMA_SCRIPT_ADDRS_ARRAY_SIZE_V4	42
 
 static void sdma_add_scripts(struct sdma_engine *sdma,
 		const struct sdma_script_start_addrs *addr)
@@ -1629,6 +1652,9 @@ static void sdma_load_firmware(const struct firmware *fw, void *context)
 		case 3:
 			sdma->script_number = SDMA_SCRIPT_ADDRS_ARRAY_SIZE_V3;
 			break;
+		case 4:
+			sdma->script_number = SDMA_SCRIPT_ADDRS_ARRAY_SIZE_V4;
+			break;
 		default:
 			dev_err(sdma->dev, "unknown firmware version\n");
 			goto err_firmware;
-- 
1.8.0

